File: 4-1.py
#!/bin/env python

import numpy

class Bingo:
    def __init__(self):
        self.board = numpy.zeros((5, 5), dtype=numpy.uint)
        self.marks = numpy.ones((5, 5), dtype=numpy.uint)

    def mark(self, num):
        for row in range(5):
            for col in range(5):
                if self.board[row, col] == num:
                    self.marks[row, col] = 0

    def won(self):
        for row in range(5):
            won = True
            for col in range(5):
                if self.marks[row, col] == 1:
                    won = False
                    break
            if won:
                return True

        for col in range(5):
            won = True
            for row in range(5):
                if self.marks[row, col] == 1:
                    won = False
                    break
            if won:
                return True

        return False

    def score(self):
        return int((self.board * self.marks).sum())

    def __str__(self):
        s = f"<{self.__class__.__name__}:\n"
        for row in range(5):
            for col in range(5):
                if self.marks[row, col] == 0:
                    s += "*"
                s += str(self.board[row, col])
                s += " "
            s += "\n"
        s += ">"
        return s


input = [line.strip() for line in open('4.input').readlines()]

draws = [int(d) for d in input.pop(0).split(",")]
boards = []

while len(input) > 0:
    input.pop(0)
    board = Bingo()
    for row in range(5):
        numbers = [int(n) for n in input.pop(0).split()]
        for col in range(5):
            board.board[row, col] = numbers[col]
    boards.append(board)

for draw in draws:
    for board in boards:
        board.mark(draw)
        if board.won():
            print(board.score() * draw)
            exit()

File: 4-1.rb
#!/usr/bin/env ruby

require 'numo/narray'

class Bingo
  attr_accessor :board
  attr_accessor :marks

  def initialize
    @board = Numo::UInt8.new(5, 5).fill 0
    @marks = Numo::UInt8.new(5, 5).fill 1
  end

  def mark(num)
    0.upto(4) do |row|
      0.upto(4) do |col|
        if @board[row, col] == num
          @marks[row, col] = 0
        end
      end
    end
  end

  def won?
    # Check each row
    0.upto(4) do |row|
      won = true
      0.upto(4) do |col|
        if @marks[row, col] == 1
          won = false
          break
        end
      end
      return true if won
    end

    # Check each column
    0.upto(4) do |col|
      won = true
      0.upto(4) do |row|
        if @marks[row, col] == 1
          won = false
          break
        end
      end
      return true if won
    end

    false
  end

  def score
    (@board * @marks).sum
  end

  def to_s
    s = "<#{self.class}:\n"
    0.upto(4) do |row|
      0.upto(4) do |col|
        s += '*' if @marks[row, col] == 0
        s += @board[row, col].to_s
        s += ' '
      end
      s += "\n"
    end
    s += '>'
    return s
  end
end

input = File.read('4.input').lines.map(&:strip)

draws = input.shift.split(',').map(&:to_i)
boards = []

until input.empty?
  input.shift  # Blank line separator
  board = Bingo.new
  0.upto(4) do |row|
    numbers = input.shift.split.map(&:to_i)
    0.upto(4) do |col|
      board.board[row, col] = numbers[col]
    end
  end
  boards.append board
end

draws.each do |draw|
  boards.each do |board|
    board.mark draw
  end

  winning = boards.index &:won?
  if winning
    print boards[winning].score * draw, "\n"
    exit
  end
end
